### PR TITLE
EstEID 2018+ driver

### PR DESCRIPTION
**Pull request by martinpaljak**
This adds card driver for Estonian eID from 2018 onwards.

Next iteration will merge pkcs15-esteid and get rid of Estonia related stuff in card-mcrd.c

### CODE REVIEW

File: src/libopensc/card-esteid2018.c [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+#include "asn1.h"
+#include "internal.h"
+#include "gp.h"
+
+static const struct sc_atr_table esteid_atrs[] = {
+    {"3b:db:96:00:80:b1:fe:45:1f:83:00:12:23:3f:53:65:49:44:0f:90:00:f1", NULL,
+     "EstEID 2018", SC_CARD_TYPE_ESTEID_2018, 0, NULL},
+    {NULL, NULL, NULL, 0, 0, NULL}};
+
+static struct sc_aid IASECC_AID = {{0xA0, 0x00, 0x00, 0x00, 0x77, 0x01, 0x08,
```
Comment (plaes) [inline src/libopensc/card-esteid2018.c]: const

Comment (plaes) [inline src/libopensc/card-esteid2018.c]: const

File: src/libopensc/card-esteid2018.c [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+                   "DF select failed");
+    } else if (pathlen == 2) {
+      LOG_TEST_RET(card->ctx, esteid_select(card, 0x02, path[0], path[1]),
+                   "EF select failed");
+
+      if (file_out != NULL) // Just make a dummy file
+      {
+        file = sc_file_new();
+        if (file == NULL)
+          LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
```
Comment (plaes) [inline src/libopensc/card-esteid2018.c]: leaking the `path` here?

File: src/libopensc/card-esteid2018.c [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+  LOG_FUNC_RETURN(ctx, apdu.resplen);
+}
+
+static int esteid_set_security_env(sc_card_t *card,
+                                   const sc_security_env_t *env, int se_num) {
+  struct esteid_priv_data *priv;
+  sc_apdu_t apdu;
+
+  unsigned char cse_crt_aut[] = {0x80, 0x04, 0xFF, 0x20, 0x08,
```
Comment (plaes) [inline src/libopensc/card-esteid2018.c]: all three `const`?

File: src/libopensc/pkcs15-esteid2018.c [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    pin_info.auth_id.value[0] = esteid_pin_authid[i];
+    pin_info.auth_type = SC_PKCS15_PIN_AUTH_TYPE_PIN;
+    pin_info.attrs.pin.reference = esteid_pin_ref[i];
+    pin_info.attrs.pin.flags = esteid_pin_flags[i];
+    pin_info.attrs.pin.type = SC_PKCS15_PIN_TYPE_ASCII_NUMERIC;
+    pin_info.attrs.pin.min_length = esteid_pin_min[i];
+    pin_info.attrs.pin.stored_length = 12;
+    pin_info.attrs.pin.max_length = 12;
+    pin_info.attrs.pin.pad_char = 0xFF;
+    pin_info.tries_left = 3;
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: Shouldn't tries_left be `-1` if it's unknown?

File: src/libopensc/pkcs15-esteid2018.c [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  if (r != 3)
+  {
+
+    LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+  }
+
+  for (i = 0; i < r; i++)
+  {
+    // trigger PIN counter refresh via pin_cmd
+    r = sc_pkcs15_get_pin_info(p15card, objs[i]);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: why is this needed? this should be called automatically if the application requests the pin info

File: src/libopensc/pkcs15-esteid2018.c [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    prkey_info.id.value[0] = prkey_pin[i];
+    prkey_info.native = 1;
+    prkey_info.key_reference = i + 1;
+    prkey_info.field_length = field_length;
+    prkey_info.usage = prkey_usage[i];
+
+    strlcpy(prkey_obj.label, prkey_name[i], sizeof(prkey_obj.label));
+    prkey_obj.auth_id.len = 1;
+    prkey_obj.auth_id.value[0] = prkey_pin[i];
+    prkey_obj.user_consent = (i == 1 ? 1 : 0);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: use `int prkey_consent[2] = {0, 1};` above

File: src/libopensc/card-esteid2018.c [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+
+  sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0xA4, p1, 0x0C);
+  if (id1 != 0x3F && id2 != 0x00) {
+    apdu.cse = SC_APDU_CASE_3_SHORT;
+    apdu.lc = 2;
+    apdu.data = sbuf;
+    apdu.datalen = 2;
+  }
+  apdu.le = 0x00;
+  apdu.resplen = 0;
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Just discovered that, somehow my notifications for this repo were turned off and I based it on 0.19.0 tag. Will update.w

File: src/libopensc/pkcs15-esteid2018.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+  p15card->tokeninfo->flags = SC_PKCS15_TOKEN_EID_COMPLIANT |
+                              SC_PKCS15_TOKEN_READONLY;
+
+  /* add certificates */
+  for (i = 0; i < 2; i++)
+  {
+    static const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+    static char const *esteid_cert_paths[2] = {"3f00adf13401", "3f00adf2341f"};
+    static int esteid_cert_ids[2] = {1, 2};
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/pkcs15-esteid2018.c]: None, to be removed.

File: src/libopensc/pkcs15-esteid2018.c (hunk 5 above)
Comment (martinpaljak) [reply src/libopensc/pkcs15-esteid2018.c]: For `pkcs15-tool -D/--list-pins`. Which could of course be changed to try to update that info when not available.

File: src/libopensc/pkcs15-esteid2018.c (hunk 4 above)
Comment (martinpaljak) [reply src/libopensc/pkcs15-esteid2018.c]: It gets updated after initial adding and 3 is the specification-defined value.

File: src/libopensc/card-esteid2018.c [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+    } else if (pathlen == 2) {
+      LOG_TEST_RET(card->ctx, esteid_select(card, 0x02, path[0], path[1]),
+                   "EF select failed");
+
+      if (file_out != NULL) // Just make a dummy file
+      {
+        file = sc_file_new();
+        if (file == NULL)
+          LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+        file->path = *in_path;
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: They don't seem to be necessary?

File: src/libopensc/card-esteid2018.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Which made me think: why not just have a changed short name like "sc_make_apdu" instead of the weird-looking _ex?

File: src/libopensc/card-esteid2018.c [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+static const struct sc_atr_table esteid_atrs[] = {
+    {"3b:db:96:00:80:b1:fe:45:1f:83:00:12:23:3f:53:65:49:44:0f:90:00:f1", NULL,
+     "EstEID 2018", SC_CARD_TYPE_ESTEID_2018, 0, NULL},
+    {NULL, NULL, NULL, 0, 0, NULL}};
+
+static struct sc_aid IASECC_AID = {{0xA0, 0x00, 0x00, 0x00, 0x77, 0x01, 0x08,
+                                    0x00, 0x07, 0x00, 0x00, 0xFE, 0x00, 0x00,
+                                    0x01, 0x00},
+                                   16};
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Yes and no.
- card-iasecc.c is indeed too bloated and complex already, this complexity is not needed for Estonian card (speed/reliability)
- card-esteid.c (once amended) would also integrate pieces from card-mcrd.c that work with pre-iasecc cards in Estonia. I gave it a long thought and decided that a separate driver is more meaningful than trying to amend or refactor card-iasecc.c

It comes down to a slim card driver that does only what is necessary (working PKCS#11) in a static setup without any metadata reading or parsing.

**Comment by martinpaljak** [comment]
> Please use tabs for indenting. Looks quite OK, nothing serious.
It has been some time since I wrote C so my setup has changed. Using VS Code with clang-format seems pretty neat. Maybe we can add necessary tooling/config into the source tree so that code formatting would be automated with a tool (like clang-format) ? I'll keep using the default formatting of VScode+clang-format plugin until the PR is non-draft or until I figure out how to configure clang-format in a suitable way.

File: src/libopensc/pkcs15-esteid2018.c (hunk 6 above)
Comment (martinpaljak) [reply src/libopensc/pkcs15-esteid2018.c]: Done

File: src/libopensc/card-esteid2018.c (hunk 3 above)
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: check

File: src/libopensc/card-esteid2018.c (hunk 1 above)
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: check

**Comment by frankmorgner** [comment]
> I'll keep using the default formatting of VScode+clang-format plugin until the PR is non-draft or until I figure out how to configure clang-format in a suitable way.

The way of achieving unified code style in OpenSC or recommendations thereof is still a todo. Sure, you can postpone a (somewhat) unified style for later.

File: src/libopensc/pkcs15-esteid2018.c [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+	set_string(&p15card->tokeninfo->label, "ID-kaart");
+	set_string(&p15card->tokeninfo->manufacturer_id, "IDEMIA");
+
+	/* Read docnr */
+	sc_format_path("3F00D003", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
+	LOG_TEST_RET(card->ctx, r, "read docnr failed");
+	const unsigned char *tag = sc_asn1_find_tag(card->ctx, buff, r, 0x04, &taglen);
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: (size_t) r

File: src/libopensc/pkcs15-esteid2018.c [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	p15card->tokeninfo->serial_number = memcpy(p15card->tokeninfo->serial_number, tag, taglen);
+	p15card->tokeninfo->serial_number[taglen] = '\0';
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_READONLY;
+
+	/* add certificates */
+	for (i = 0; i < 2; i++) {
+		const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *esteid_cert_paths[2] = {"3f00:adf1:3401", "3f00:adf2:341f"};
+		const int esteid_cert_ids[2] = {1, 2};
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: const u8 esteid_cert_ids

File: src/libopensc/pkcs15-esteid2018.c [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+		free(cn_name);
+		sc_pkcs15_free_certificate(cert);
+	}
+
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		static const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
+		static const int esteid_pin_min[3] = {4, 5, 8};
+		static const int esteid_pin_ref[3] = {0x01, 0x85, 0x02};
+		static const int esteid_pin_authid[3] = {1, 2, 3};
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: static const u8 esteid_pin_authid

File: src/libopensc/pkcs15-esteid2018.c [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+		static const char *esteid_pin_path[3] = {"3F00", "3F00ADF2", "3F00"};
+
+		static const int esteid_pin_flags[3] = {
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: static const unsigned int esteid_pin_flags

File: src/libopensc/pkcs15-esteid2018.c [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+			}
+		}
+		free(cn_name);
+		sc_pkcs15_free_certificate(cert);
+	}
+
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		static const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
+		static const int esteid_pin_min[3] = {4, 5, 8};
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: static const size_t esteid_pin_min

File: src/libopensc/pkcs15-esteid2018.c [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+	// trigger PIN counter refresh via pin_cmd
+	for (i = 0; i < r; i++) {
+		r = sc_pkcs15_get_pin_info(p15card, objs[i]);
+		if (r < 0)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+
+	/* add private keys */
+	for (i = 0; i < 2; i++) {
+		int prkey_pin[2] = {1, 2};
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: const u8 prkey_pin

File: src/libopensc/pkcs15-esteid2018.c [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+
+		const char *prkey_name[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *prkey_path[2] = {"3F00:ADF1", "3F00:ADF2"};
+		const int prkey_usage[2] = {SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_DERIVE,
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: const unsigned int prkey_usage

File: src/libopensc/card-esteid2018.c [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "SELECT failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int esteid_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out) {
+	unsigned char pathbuf[SC_MAX_PATH_SIZE], *path = pathbuf;
+	int pathlen;
```
Comment (metsma) [inline src/libopensc/card-esteid2018.c]: size_t pathlen;

File: src/libopensc/card-esteid2018.c [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+	apdu.data = sbuf;
+	apdu.datalen = datalen;
+	apdu.le = MIN(256, outlen);
+	apdu.resp = out;
+	apdu.resplen = outlen;
+
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "PSO CDS/INTERNAL AUTHENTICATE failed");
+
+	LOG_FUNC_RETURN(card->ctx, apdu.resplen);
```
Comment (metsma) [inline src/libopensc/card-esteid2018.c]: (int) apdu.resplen

File: src/libopensc/card-esteid2018.c [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+		pathlen -= 2;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+// temporary hack, overload 6B00 SW processing
+static int esteid_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t count, unsigned long flags) {
+	int r;
+	LOG_FUNC_CALLED(card->ctx);
+	void *saved = card->ops->check_sw;
```
Comment (metsma) [inline src/libopensc/card-esteid2018.c]: typedef int (*check_sw)(struct sc_card *card,unsigned int sw1,unsigned int sw2);
	check_sw saved = card->ops->check_sw;

Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Correct me if I'm wrong, but handling a void pointer in this context should be sufficient effectively (a pointer is a pointer no matter where it points)

Comment (metsma) [reply src/libopensc/card-esteid2018.c]: https://stackoverflow.com/questions/36645660/why-cant-i-cast-a-function-pointer-to-void

File: src/libopensc/card-esteid2018.c [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+
+	memcpy(path, in_path->value, in_path->len);
+	pathlen = in_path->len;
+
+	while (pathlen >= 2) {
+		if (memcmp(path, "\x3F\x00", 2) == 0) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "MF select failed");
+		} else if (pathlen >= 2 && path[0] == 0xAD) {
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: pathlen will be at least 2, due to the `while` condition

File: src/libopensc/card-esteid2018.c [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+	priv = DRVDATA(card);
+	priv->sec_env = *env;
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int esteid_compute_signature(sc_card_t *card, const u8 *data, size_t datalen, u8 *out, size_t outlen) {
+	struct esteid_priv_data *priv = DRVDATA(card);
+	struct sc_security_env *env = NULL;
+	struct sc_apdu apdu;
+	u8 sbuf[SIGNATURE_PAYLOAD_SIZE] = {0};
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: Either use `= {0};` or `memset(sbuf, 0x00, SIGNATURE_PAYLOAD_SIZE - datalen);`, but don't use both. The former should work starting with C99.

File: src/libopensc/card-esteid2018.c [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+
+	env = &priv->sec_env;
+	// left-pad if necessary
+	memcpy(&sbuf[SIGNATURE_PAYLOAD_SIZE - datalen], data, MIN(datalen, SIGNATURE_PAYLOAD_SIZE));
+	memset(sbuf, 0x00, SIGNATURE_PAYLOAD_SIZE - datalen);
+	datalen = SIGNATURE_PAYLOAD_SIZE;
+
+	switch (env->key_ref[0]) {
+	case 1: /* authentication key */
+		sc_format_apdu_ex(card, &apdu, 0x88, 0, 0, sbuf, datalen, out, MIN(256, outlen));
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: Do you want to use `SC_MAX_APDU_RESP_SIZE` instead of `256`?

File: src/libopensc/card-esteid2018.c [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+		LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "Cannot select MF");
+	} else if (pin_reference == 0x85) {
+		LOG_TEST_RET(card->ctx, esteid_select_file(card, &adf2, NULL), "Cannot select QSCD AID");
+	} else {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+
+	get_pin_info[6] = pin_reference & 0x0F; // mask out local/global
+	// XXX: handling of default Le of 0x00 could be easier
+	sc_format_apdu_ex(card, &apdu, 0xCB, 0x3F, 0xFF, get_pin_info, sizeof(get_pin_info), apdu_resp, MIN(256, sizeof(apdu_resp)));
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: Use `SC_MAX_APDU_RESP_SIZE` for declaring `apdu_resp` instead of `SC_MAX_APDU_BUFFER_SIZE` and you're good to specify the length with `sizeof(apdu_resp)` only

File: src/libopensc/card-esteid2018.c [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+	SC_TRANSMIT_TEST_RET(card, apdu, "GET DATA(pin info) failed");
+
+	// XXX: sc_asn1_find_tag with the following payload (to get to tag 0x9B):
+	// https://lapo.it/asn1js/#cB6_gQEaoBiaAQObAQOhEIwG8wAAc0MAnAbzAABzQwA
+	return apdu_resp[13];
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: Yes, parsing the content is the way to go! At the very least, you should check whether you've received at least 14 bytes.

Comment (frankmorgner) [reply src/libopensc/card-esteid2018.c]: cast to int?

File: src/libopensc/card-esteid2018.c [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+	} else if (data->cmd == SC_PIN_CMD_UNBLOCK) {
+		// Verify PUK, then issue UNBLOCK
+		// VERIFY
+		memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
+		tmp.cmd = SC_PIN_CMD_VERIFY;
+		tmp.pin_reference = 0x02; // hardcoded, ugly
+		tmp.pin2.len = 0;
+		r = iso_ops->pin_cmd(card, &tmp, tries_left);
+		sc_mem_clear(&tmp, sizeof(tmp));
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: no need to clear the memory, because the actual PINs (`tmp.pin1.data`/`tmp.pin2.data`) are only pointers.

File: src/libopensc/card-esteid2018.c [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+		LOG_TEST_RET(card->ctx, r, "GET DATA(pin info) failed");
+
+		data->pin1.tries_left = r;
+		data->pin1.max_tries = -1; // "no support, which means the one set in PKCS#15 emulation sticks
+		data->pin1.logged_in = SC_PIN_STATE_UNKNOWN;
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+	} else if (data->cmd == SC_PIN_CMD_UNBLOCK) {
+		// Verify PUK, then issue UNBLOCK
+		// VERIFY
+		memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: why don't you use `tmp = *data`?

File: src/libopensc/card-esteid2018.c [hunk 28]
```diff
@@ -0,0 +1,327 @@
...
+		tmp.cmd = SC_PIN_CMD_VERIFY;
+		tmp.pin_reference = 0x02; // hardcoded, ugly
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: use a human readable define instead of `0x02`.

File: src/libopensc/iso7816.c [hunk 29]
```diff
@@ -152,10 +152,8 @@ iso7816_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t coun
 	fixup_transceive_length(card, &apdu);
 	r = sc_transmit_apdu(card, &apdu);
 	LOG_TEST_RET(ctx, r, "APDU transmit failed");
-	if (apdu.resplen == 0)
-		LOG_FUNC_RETURN(ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
 
-	r =  sc_check_sw(card, apdu.sw1, apdu.sw2);
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
```
Comment (frankmorgner) [inline src/libopensc/iso7816.c]: This change looks OK.

With this applied, do you still need the workaround with `esteid_check_sw`?

File: src/libopensc/pkcs15-esteid2018.c [hunk 30]
```diff
@@ -0,0 +1,220 @@
...
+
+static void set_string(char **strp, const char *value) {
+	if (*strp)
+		free(*strp);
+	*strp = value ? strdup(value) : NULL;
+}
+
+static int sc_pkcs15emu_esteid2018_init(sc_pkcs15_card_t *p15card) {
+	sc_card_t *card = p15card->card;
+	u8 buff[128];
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: looking at your code below, 11 bytes should be enough for buff

File: src/libopensc/pkcs15-esteid2018.c [hunk 31]
```diff
@@ -0,0 +1,220 @@
...
+	size_t field_length = 0, taglen, j;
+	sc_path_t tmppath;
+
+	set_string(&p15card->tokeninfo->label, "ID-kaart");
+	set_string(&p15card->tokeninfo->manufacturer_id, "IDEMIA");
+
+	/* Read docnr */
+	sc_format_path("3F00D003", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: sizeof buff instead of 11?

File: src/libopensc/pkcs15-esteid2018.c [hunk 32]
```diff
@@ -0,0 +1,220 @@
...
+				free(token_name);
+			}
+		}
+		free(cn_name);
+		sc_pkcs15_free_certificate(cert);
+	}
+
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-esteid2018.c]: aren't there better names for PIN1 and PIN2?

Comment (metsma) [reply src/libopensc/pkcs15-esteid2018.c]: This is how they are labeled in PIN envelope when they are handed to citizens

Comment (frankmorgner) [reply src/libopensc/pkcs15-esteid2018.c]: Then the *real* fix is needed elsewhere...

File: src/libopensc/card-esteid2018.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (frankmorgner) [reply src/libopensc/card-esteid2018.c]: adding _ex lets developers know which one is the successor. Introducing new names adds confusion. Ideally, I'd like to remove the function without the _ex, but I'm too lazy to go through all the codebase...

File: src/libopensc/card-esteid2018.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (frankmorgner) [reply src/libopensc/card-esteid2018.c]: well, the iso7816 driver is quite successful in offering a generic and extensible implementation. For ias/ecc the same should be possible. Unfortunately, I think it's already too late...

File: src/libopensc/card-esteid2018.c [hunk 33]
```diff
@@ -0,0 +1,411 @@
...
+    data->pin1.max_tries =
+        -1; // "no support, which means the one set in PKCS#15 emulation sticks
+    data->pin1.logged_in = SC_PIN_STATE_UNKNOWN;
+    return SC_SUCCESS;
+  } else if (data->cmd == SC_PIN_CMD_UNBLOCK) {
+    // Verify PUK, then issue "CHANGE"
+    // VERIFY
+    memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
+    tmp.cmd = SC_PIN_CMD_VERIFY;
+    tmp.pin_reference = 0x02; // hardcoded, ugly
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: What do you mean?

File: src/libopensc/card-esteid2018.c [hunk 34]
```diff
@@ -0,0 +1,411 @@
...
+    tmp.pin2.len = 0;
+    r = iso_ops->pin_cmd(card, &tmp, tries_left);
+    sc_mem_clear(&tmp, sizeof(tmp));
+    LOG_TEST_RET(card->ctx, r, "VERIFY during unblock failed");
+
+    if (data->pin_reference == 0x85) {
+      LOG_TEST_RET(card->ctx, esteid_select_file(card, &adf2, NULL),
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: You mean the caller?

File: src/libopensc/iso7816.c (hunk 29 above)
Comment (martinpaljak) [reply src/libopensc/iso7816.c]: Yes.

**Comment by frankmorgner** [comment]
please also look at the two unresolved comments above

File: src/libopensc/pkcs15-esteid2018.c [hunk 35]
```diff
@@ -0,0 +1,220 @@
...
+		r = sc_pkcs15emu_add_x509_cert(p15card, &cert_obj, &cert_info);
+		if (r < 0)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+		// Read data from first cert
+		if (i != 0)
+			continue;
+
+		sc_pkcs15_cert_t *cert = NULL;
+		r = sc_pkcs15_read_certificate(p15card, &cert_info, &cert);
```
Comment (metsma) [inline src/libopensc/pkcs15-esteid2018.c]: Fails to read Test card certificate. Reads only 1024 bytes.
Seems like the size is hardcoded here
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15.c#L2355

File: src/libopensc/card-esteid2018.c (hunk 21 above)
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Indeed

File: src/libopensc/pkcs15-esteid2018.c (hunk 35 above)
Comment (martinpaljak) [reply src/libopensc/pkcs15-esteid2018.c]: Indeed. fixed this with latest push to this branch. But the right thing to do would be to remove that fixed 1k buffer from core so that the file could be read "until fully read" irrelevant of size.

File: src/libopensc/card-esteid2018.c (hunk 25 above)
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Check

File: src/libopensc/card-esteid2018.c [hunk 36]
```diff
@@ -0,0 +1,330 @@
...
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+
+	get_pin_info[6] = pin_reference & 0x0F; // mask out local/global
+	sc_format_apdu_ex(card, &apdu, 0xCB, 0x3F, 0xFF, get_pin_info, sizeof(get_pin_info), apdu_resp, sizeof(apdu_resp));
+	SC_TRANSMIT_TEST_RET(card, apdu, "GET DATA(pin info) failed");
+
+	// XXX: sc_asn1_find_tag with the following payload (to get to tag 0x9B):
+	// https://lapo.it/asn1js/#cB6_gQEaoBiaAQObAQOhEIwG8wAAc0MAnAbzAABzQwA
+	return (int)apdu_resp[13];
```
Comment (frankmorgner) [inline src/libopensc/card-esteid2018.c]: need to check response size

**Comment by martinpaljak** [comment]
> See comment above. There is also [this remark](https://github.com/OpenSC/OpenSC/pull/1635/files#r268146297). Please exclude `.clang-format` from this PR.
> 
> The rest looks good.

Addressed both, removed clang-format, rebased on master.

File: src/libopensc/card-esteid2018.c [hunk 37]
```diff
@@ -0,0 +1,333 @@
...
+
+#define SC_TRANSMIT_TEST_RET(card, apdu, text)                                                                                                      \
+	do {                                                                                                                               \
+		LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");                                            \
+		LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), text);                                                      \
+	} while (0)
+
+static int esteid_match_card(sc_card_t *card) {
+	int i = 0;
+	int r = 0;
```
[the remark this answers is not in the document]
Comment (martinpaljak) [reply src/libopensc/card-esteid2018.c]: Indeed, reduced in last push, that rebases it against master and reduces number of "review updates" commits.
